wknml/nml_generation.py

```diff
@@ -0,0 +1,182 @@
+
+  return nml
+
+
+def generate_graph(nml: NML) -> Tuple[Dict[str, List[nx.Graph]], Dict]:
+    nml = nml._replace(groups=discard_children_hierarchy(nml.groups))
+    group_dict = {}
+    for group in nml.groups:
+        graphs_in_current_group = []
+        for tree in nml.trees:
+            if tree.groupId == group.id:
+                graphs_in_current_group.append(nml_tree_to_graph(tree))
+        group_dict[group.name] = graphs_in_current_group
+
+    nml_parameters = nml.parameters
+    parameter_dict = {}
+
+    parameter_list = ["name", "scale", "offset", "time", "editPosition", "editRotation", "zoomLevel",
+                      "taskBoundingBox", "userBoundingBox"]
+
+    for parameter in parameter_list:
+        if getattr(nml_parameters, parameter) is not None:
+            parameter_dict[parameter] = getattr(nml_parameters, parameter)
+
+    for comment in nml.comments:
+        for group in group_dict.values():
+            for tree in group:
+                if comment.node in tree.nodes:
+                    tree.nodes[comment.node]["comment"] = comment.content
+
+    for branchpoint in nml.branchpoints:
+        for group in group_dict.values():
+            for tree in group:
+                if branchpoint.id in tree.nodes:
+                    tree.nodes[branchpoint.id]["branchpoint"] = branchpoint.time
+
+    return group_dict, parameter_dict
+
+
+def nml_tree_to_graph(tree: Tree) -> nx.Graph:
+    optional_attribute_list = ["rotation", "inVp", "inMag", "bitDepth", "interpolation"]
+
+    graph = nx.Graph(id= tree.id, color=tree.color, name=tree.name, groupId=tree.groupId)
+    for node in tree.nodes:
+        node_id = node.id
+        graph.add_node(node_id, id=node_id, radius=node.radius, position=node.position)
+        for optional_attribute in optional_attribute_list:
+            if getattr(node, optional_attribute) is not None:
+                graph.nodes[node_id][optional_attribute] = getattr(node, optional_attribute)
+
+    graph.add_edges_from([(edge.source, edge.target) for edge in tree.edges])
+
+    return graph
+
+
+def extract_nodes_and_edges_from_graph(graph: nx.Graph) -> Tuple[List[Node], List[Edge]]:
+    node_nml = [Node(id=graph.nodes[node]["id"],
+                   position=graph.nodes[node]["position"],
+                   radius=graph.nodes[node]["radius"] if "radius" in graph.nodes[node] else 1.0,
+                   rotation=graph.nodes[node]["rotation"] if "rotation" in graph.nodes[node] else None,
+                   inVp=graph.nodes[node]["inVp"] if "inVp" in graph.nodes[node] else None,
+                   inMag=graph.nodes[node]["inMag"] if "inMag" in graph.nodes[node] else None,
+                   bitDepth=graph.nodes[node]["bitDepth"] if "bitDepth" in graph.nodes[node] else None,
+                   interpolation=graph.nodes[node]["interpolation"] if "interpolation" in graph.nodes[node] else None,
+                   time=graph.nodes[node]["time"] if "time" in graph.nodes[node] else 0)
```

Comment:
The default should be `None` so that there is no time persisted at all.
It is important that the lib supports *not* persisting optional node attributes, because there might be many nodes and we don't want the NMLs to be unnecessarily large.

Reply:
ok :)